mypy/checker.py

```diff
@@ -3870,6 +3870,27 @@ def check_simple_assignment(
             rvalue_type = self.expr_checker.accept(
                 rvalue, lvalue_type, always_allow_any=always_allow_any
             )
+            if isinstance(get_proper_type(lvalue_type), UnionType):
+                # Try re-inferring r.h.s. in empty context, and use that if it
+                # results in a narrower type. We don't do this always because this
+                # may cause some perf impact, plus we want to partially preserve
+                # the old behavior. This helps with various practical examples, see
+                # e.g. testOptionalTypeNarrowedByGenericCall.
+                with self.msg.filter_errors() as local_errors, self.local_type_map() as type_map:
+                    alt_rvalue_type = self.expr_checker.accept(
+                        rvalue, None, always_allow_any=always_allow_any
+                    )
+                if (
+                    not local_errors.has_new_errors()
+                    # Skip literal types, as they have special logic (for better errors).
+                    and not isinstance(get_proper_type(rvalue_type), LiteralType)
```

Comment:
nit: seems like we can lift out `and not isinstance(get_proper_type(rvalue_type), LiteralType)` as a small perf optimisation